packages/jsii/lib/assembler.ts

```
@@ -0,0 +1,879 @@
+            if (!symbol) { continue; }
+            for (const node of this._typeChecker.getExportsOfModule(symbol)) {
+                await this._visitNode(node.declarations[0]);
+            }
+        }
+        for (const deferred of this._deferred) {
+            deferred();
+        }
+
+        // Skip emitting if any diagnostic message is an error
+        if (this._diagnostics.find(diag => diag.category === ts.DiagnosticCategory.Error) != null) {
+            LOG.debug('Skipping emit due to errors.');
+            // Clearing ``this._types`` to allow contents to be garbage-collected.
+            delete this._types;
+            try {
+                return { diagnostics: this._diagnostics, emitSkipped: true };
+            } finally {
+                // Clearing ``this._diagnostics`` and ``this._deferred`` to allow contents to be garbage-collected.
+                delete this._diagnostics;
+                delete this._deferred;
+            }
+        }
+
+        const assembly = {
+            schema: spec.SchemaVersion.V1_0,
+            name: this.projectInfo.name,
+            version: this.projectInfo.version,
+            description: this.projectInfo.description || this.projectInfo.name,
+            license: this.projectInfo.license,
+            homepage: this.projectInfo.homepage || this.projectInfo.repository.url,
+            author: this.projectInfo.author,
+            contributors: this.projectInfo.contributors && [...this.projectInfo.contributors],
+            repository: this.projectInfo.repository,
+            dependencies: _toDependencies(this.projectInfo.dependencies),
+            bundled: this.projectInfo.bundleDependencies,
+            types: this._types,
+            targets: this.projectInfo.targets,
+            readme,
+            fingerprint: '<TBD>'
+        };
+
+        const validator = new Validator(this.projectInfo, assembly);
+        const validationResult = await validator.emit();
```

What's the value in making the validator an emitter? What does it emit really? What's the polymorphic benefit?

---

The validator actually further annotates the type specification (for example, it tags the`overridden` members).

---

That sounds wrong.

---

Yah I think ultimately I want to merge those into the `Assembler` class entirely. Separating made it easier for me to wrap my head around the two classes of concerns I had to deal with here:
1. Understanding the type model that I obtain from the `tsc` API
2. Make sure we generate valid type definitions for JSII standards

But we can "fail faster" if we merge...